### PR TITLE
Even faster builds: CI run under 8 minutes.

### PR DESCRIPTION
**Pull request by MichalCz**
This fixes build scripts and fixes a couple of flaky tests. Also removes the ones I couldn't fix.

- Faster build
- Dependencies update
- Better job structure
- Jobs separated to own files

Prominent changes:

1. The `build-all.js` script calls webpack and installs packages in `dist/` - this cut down over 1 minute of build time.
2. Therefore running `prepack` in packages is not needed (so no `lerna` calls).
3. Refapps build is building all apps in one build script
4. Packaging apps is still using Lerna (so some improvement could be there)

The build scripts:

1. The workflows are separated as reusable workflows with arguments,
2. The matrix for node 14 and 16 was a bit of a slowdown, so these two got separated,
3. The files prefixed with `_main_` now are the ones that run on `pr`.

### CODE REVIEW

**Review by a-tylenda — APPROVED** [review]

**Review by daro1337 — APPROVED** [review]
LGTM

**Comment by jan-warchol** [comment]
Are the changes in `.github/workflows/`, `bdd/` and `packages/` independent? Any chance we could split this PR easily?

File: package.json [hunk 1]
```diff
@@ -19,27 +19,27 @@
     "build": "yarn build:all",
     "build:packages-esbuild": "lerna run build:esbuild",
     "build:all": "yarn build:packages && yarn build:refapps && yarn build:docker",
-    "prebuild:refapps": "scripts/build-all.js --dirs packages/reference-apps/* --config-name=tsconfig.json",
-    "build:refapps": "LOCAL_PACKAGES=true lerna run build:refapps:only",
     "postbuild:refapps": "LOCAL_PACKAGES=true lerna run postbuild:refapps",
-    "prebuild:packages": "scripts/build-all.js --dirs packages/* --config-name=tsconfig.build.json",
+    "prebuild:packages": "scripts/build-all.js --dirs packages/* --config-name=tsconfig.build.json --copy-dist",
+    "prebuild:refapps:node": "NO_WORKSPACE=true scripts/build-all.js --dirs packages/reference-apps/* --config-name=tsconfig.json --copy-dist",
```
Comment (jan-warchol) [inline package.json:24]: This one seems unused.

Comment (jan-warchol) [reply package.json:24]: wrong. nevermind.

File: .github/workflows/build-docker-prerunner.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
+name: Reusable workflow example
```
Comment (jan-warchol) [inline .github/workflows/build-docker-prerunner.yml:1]: Can we update this name?

File: .github/workflows/build-docker-runner-node.yml [hunk 3]
```diff
@@ -0,0 +1,40 @@
+name: Reusable workflow example
```
Comment (jan-warchol) [inline .github/workflows/build-docker-runner-node.yml:1]: Please update this name.

File: .github/workflows/build-docker-runner-python.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
+name: Reusable workflow example
```
Comment (jan-warchol) [inline .github/workflows/build-docker-runner-python.yml:1]: Please update this name.

File: .github/workflows/build-docker-sth.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
+name: Reusable workflow example
```
Comment (jan-warchol) [inline .github/workflows/build-docker-sth.yml:1]: Please update this name.

File: .github/workflows/build-refapps.yml [hunk 6]
```diff
@@ -0,0 +1,39 @@
+name: Reusable workflow example
```
Comment (jan-warchol) [inline .github/workflows/build-refapps.yml:1]: Please update this name.

File: .github/workflows/build-sth.yml [hunk 7]
```diff
@@ -0,0 +1,44 @@
+name: Reusable workflow example
```
Comment (jan-warchol) [inline .github/workflows/build-sth.yml:1]: Please update this name.

File: .github/workflows/preinstall-deps.yml [hunk 8]
```diff
@@ -0,0 +1,33 @@
+name: Analyze Source Code
```
Comment (jan-warchol) [inline .github/workflows/preinstall-deps.yml:1]: This name seems wrong?

File: .github/workflows/test-bdd-sth-docker.yml [hunk 9]
```diff
@@ -0,0 +1,68 @@
+name: Reusable workflow example
```
Comment (jan-warchol) [inline .github/workflows/test-bdd-sth-docker.yml:1]: Please update this name.

File: .github/workflows/test-bdd-sth-no-docker.yml [hunk 10]
```diff
@@ -0,0 +1,42 @@
+name: Reusable workflow example
```
Comment (jan-warchol) [inline .github/workflows/test-bdd-sth-no-docker.yml:1]: Please update this name.

File: .github/workflows/test-bdd-sth-no-host-no-docker.yml [hunk 11]
```diff
@@ -0,0 +1,44 @@
+name: Reusable workflow example
```
Comment (jan-warchol) [inline .github/workflows/test-bdd-sth-no-host-no-docker.yml:1]: Please update this name.

File: .github/workflows/test-bdd-sth-no-host.yml [hunk 12]
```diff
@@ -0,0 +1,68 @@
+name: Reusable workflow example
```
Comment (jan-warchol) [inline .github/workflows/test-bdd-sth-no-host.yml:1]: Please update this name.

File: .github/workflows/test-bdd-sth-python.yml [hunk 13]
```diff
@@ -0,0 +1,71 @@
+name: Reusable workflow example
```
Comment (jan-warchol) [inline .github/workflows/test-bdd-sth-python.yml:1]: Please update this name.

File: .github/workflows/test-unit.yml [hunk 14]
```diff
@@ -0,0 +1,40 @@
+name: Reusable workflow example
```
Comment (jan-warchol) [inline .github/workflows/test-unit.yml:1]: Please update this name.

**Review by jan-warchol — CHANGES_REQUESTED** [review]
I found some confusing github workflow names.

**Review by patuwwy — APPROVED** [review]
Massive

File: .github/workflows/build-docker-prerunner.yml (hunk 2 above)
Comment (MichalCz) [reply .github/workflows/build-docker-prerunner.yml:1]: Yes, can you help. :)

**Comment by jan-warchol** [comment]
CPM doesn't build with STH submodule pointing to this branch. See error at https://github.com/scramjetorg/platform-tests/runs/6353053124?check_suite_focus=true

**Comment by patuwwy** [comment]
> CPM doesn't build with STH submodule pointing to this branch. See error at https://github.com/scramjetorg/platform-tests/runs/6353053124?check_suite_focus=true

Sure it will fail. As now, CPM has copy of scripts, and the build flow is the same as on transform-hub `devel` branch. changes introduced in this PR make CPM can't run scripts because they don't exist anymore.

**Review by jan-warchol — CHANGES_REQUESTED** [review]
Please split this PR. As far as I understand, there are 2 or 3 separate changes here:
1. reorganization of github workflows
2. changed build system (perhaps could be split into a few smaller parts?)
3. (?) changes in bdd tests

Part 2. currently makes CPM not build, so this cannot be merged now. The PR for making CPM work is still in progress, and it's hard to say when it will be done. Because of that the other changes from this PR are blocked. So, I think it should be split.

File: .github/workflows/preinstall-deps.yml [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+      - name: Cache node_modules
+        uses: actions/cache@v3
+        with:
+          path: |
+            node_modules
+          key: ${{ runner.os }}-${{ inputs.node-version }}-${{ hashFiles('**/yarn.lock') }}
+
+      - name: Install dependencies
+        run: yarn install --ignore-engines --frozen-lockfile
```
Comment (jan-warchol) [inline .github/workflows/preinstall-deps.yml:33]: I couldn't find any references to this workflow. What is it used for?

**Review by jan-warchol — COMMENTED** [review]
A few questions about places I don't understand.

File: .github/workflows/preinstall-deps.yml (hunk 15 above)
Comment (MichalCz) [reply .github/workflows/preinstall-deps.yml:33]: I used it but it didn't help. I'd like to use it later for accelerating the dependency install, but for that I don't want to extend the complexity. Future code.

File: packages/sth/Dockerfile [hunk 16]
```diff
@@ -41,4 +38,4 @@ COPY --from=builder /build/dist ./dist
 
 RUN npm install -g ./dist/sth
 
-CMD [ "scramjet-transform-hub" ]
+CMD [ "scramjet-transform-hub", "--runtime-adapter=process" ]
```
[the remark this answers is not in the document]
Comment (MichalCz) [reply packages/sth/Dockerfile:41]: Let me review this.

[the remark this answers is not in the document]
Comment (MichalCz) [reply packages/sth/Dockerfile:41]: I'll revert this.

**Pull request merged**